ext/standard/array.c

```diff
@@ -2208,43 +2208,77 @@ PHP_FUNCTION(range)
 			zend_hash_next_index_insert_new(Z_ARRVAL_P(return_value), &tmp);
 		}
 	} else {
-		double low, high;
-		zend_long lstep;
+		zend_long low, high, lstep, __calc_size;
 long_str:
-		low = zval_get_double(zlow);
-		high = zval_get_double(zhigh);
+		low = zval_get_long(zlow);
+		high = zval_get_long(zhigh);
 		lstep = (zend_long) step;
 
 		Z_TYPE_INFO(tmp) = IS_LONG;
 		if (low > high) { 		/* Negative steps */
-			if (low - high < lstep || lstep <= 0) {
+			/* overflow may occur if high = ZEND_LONG_MIN and low = 0 */
+			if ((high != ZEND_LONG_MIN && low != 0) && (low - high < lstep || lstep <= 0)) {
 				err = 1;
 				goto err;
 			}
 
-			RANGE_CHECK_INIT_ARRAY(low, high);
+			__calc_size = ((low - high) / lstep) + 1;
+
+			if ((low == ZEND_LONG_MAX && high < 1) /* overflow check */
+				|| (high == ZEND_LONG_MIN && low > -2) /* overflow check */
+				|| __calc_size > HT_MAX_SIZE /* the array size is too big */
+			) {
+				php_error_docref(NULL, E_WARNING, "The supplied range exceeds the maximum array size: start=%ld end=%ld", high, low);
```

Comment:
An appropriate sprintf format should be used for zend_long, please see UPGRADING INTERNALS.